### PR TITLE
Streaming hoard

### PR DESCRIPTION
**Pull request by silasdavis**

### CODE REVIEW

File: streaming.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+	return consumePlaintext(accum.Plaintext, chunk.Plaintext)
+}
+
+func consumePlaintext(accum, chunk *api.Plaintext) error {
+	if chunk == nil {
+		return nil
+	}
+	if len(chunk.Salt) > 0 {
+		if len(accum.Salt) > 0 {
+			return fmt.Errorf("recevied multiple salts but there can be at most one")
```
Comment (gregdhill) [inline streaming.go]:
```suggestion
			return fmt.Errorf("received multiple salts but there can be at most one")
```

File: streaming.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		}
+		if len(accum.Data) > 0 {
+			return fmt.Errorf("recevied salt after data but salt must come before all data chunks")
```
Comment (gregdhill) [inline streaming.go]:
```suggestion
			return fmt.Errorf("received salt after data but salt must come before all data chunks")
```

File: streaming.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+		accum.GrantSpec = chunk.GrantSpec
+	}
+	return consumePlaintext(accum.Plaintext, chunk.Plaintext)
+}
+
+func consumePlaintext(accum, chunk *api.Plaintext) error {
+	if chunk == nil {
+		return nil
+	}
+	if len(chunk.Salt) > 0 {
```
Comment (gregdhill) [inline streaming.go]: Why do we need to enforce these constraints on the salt?

File: document.go [hunk 4]
```diff
@@ -7,12 +7,12 @@ import (
 	"fmt"
 
 	"github.com/gogo/protobuf/proto"
+	"github.com/monax/hoard/v6/api"
```
Comment (gregdhill) [inline document.go]: I think all hoard imports should point to v7 now...

File: go.mod [hunk 5]
```diff
@@ -2,32 +2,28 @@ module github.com/monax/hoard/v6
```
Comment (gregdhill) [inline go.mod]: This should point to v7.

File: service.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
+package hoard
+
+import (
+	"context"
+
+	"github.com/monax/hoard/v6/api"
```
Comment (gregdhill) [inline service.go]: v++

File: service_test.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
+package hoard
+
+import (
+	"context"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+	"github.com/monax/hoard/v6/api"
```
Comment (gregdhill) [inline service_test.go]: v++

File: streaming.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
+package hoard
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/monax/hoard/v6/api"
```
Comment (gregdhill) [inline streaming.go]: v++

File: streaming.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+			if err == io.EOF {
+				return &meta.Document{Meta: accum.Meta, Data: accum.Plaintext.Data}, nil
+			}
+
+			return nil, err
+		}
+
+		if d.Meta != nil {
+			if accum.Meta != nil {
+				return nil, fmt.Errorf("recevied multiple document meta but there can be at most one")
```
Comment (gregdhill) [inline streaming.go]:
```suggestion
				return nil, fmt.Errorf("received multiple document meta but there can be at most one")
```

Comment (gregdhill) [reply streaming.go]: Why can we only consume one?

File: streaming.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+	}
+}
+
+func consumePlaintextAndGrantSpec(accum, chunk *api.PlaintextAndGrantSpec) error {
+	if chunk == nil {
+		return nil
+	}
+	if chunk.GrantSpec != nil {
+		if accum.GrantSpec != nil {
+			return fmt.Errorf("recevied multiple grant specs but there can be at most one")
```
Comment (gregdhill) [inline streaming.go]:
```suggestion
			return fmt.Errorf("received multiple grant specs but there can be at most one")
```

File: streaming.go [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+		}
+		if len(accum.Plaintext.Data) > 0 {
+			return fmt.Errorf("recevied grant spec after data but spec must come before all data chunks")
```
Comment (gregdhill) [inline streaming.go]:
```suggestion
			return fmt.Errorf("received grant spec after data but spec must come before all data chunks")
```

File: streaming_test.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
+package hoard
+
+import (
+	"io"
+	"testing"
+
+	"github.com/monax/hoard/v6/api"
```
Comment (gregdhill) [inline streaming_test.go]: v++

File: streaming.go (hunk 9 above)
Comment (silasdavis) [reply streaming.go]: If we add two - which one is the real one?

Or are you consider we allow them to be composited - this seems like an unnecessary complication. Almost by definition we don't expect our metadata to be so large - much better if we can be sure once we have read the metadata we have all of it.

**Comment by silasdavis** [comment]
> Without `oneof`, I question why we need to enforce message ordering server side. In the case of the salt (which is just another byte slice) why can't we simply append?

The primary reason is that we want to be able to read the beginning of a stream sequentially and get the metadata without reading further in. If we allow out-of-order transmission then either we need to re-order it which is a pain or we loose the ability to take the HEAD of a stream and extract the metadata.

We could append on the salt but again we provide flexibility but there is no good reason to split up a salt - it is probably a mistake.

Generally having stronger invariants is better than have more arbitrary flexibility if it is not clear what good that does. We basically have some descriptive metadata and some stream. If it wasn't for the way plaintexts are handled I'd be tempted to have a message like:

```
message Document {
   Metadata Data = 1;
   bytes Data = 2;
}
``` 
And include the salt, metadata, grantspec and whatever other stuff you are only meant to say once in there. We could still go this route but I didn't want to make too many changes.

> Might be worth adding some documentation on using the js lib, but that is low priority.
I have put a far few comments in the code - I'm not sure I want to maintain more detached documentation at this point...

**Pull request merged**